### ISSUE TITLE
BBC: v5.27.8-307-g58e641fba5 -Dusethreads -DEBUGGING=both breaks Tk-Clock test files

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#17797]
Originally reported by Andreas Koenig in https://rt.cpan.org/Ticket/Display.html?id=132669.

When perl is compiled with threading and `-DEBUGGING=both`, it breaks two files in the test suite for CPAN distribution Tk-Clock.

Bisecting with:
```
$ perl Porting/bisect.pl \
-Duseithreads -DEBUGGING=both \
--start=9652c19c0aaaaa98cda620d4bf62d66217d0b4c1 \
--end=v5.27.9 \
--module=Tk::Clock
```
... points to:

```
commit 58e641fba50d93bcd84c0dc042b36d201958293d
Author: Karl Williamson <khw@cpan.org>
Date:   Thu Feb 15 22:44:24 2018 -0700
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Sun Feb 18 15:44:23 2018 -0700

    Add switch_to_globale_locale()
    
    This new API function is for use in applications that call alien library
    routines that are expecting the old pre-POSIX 2008 locale functionality,
    namely a single global locale accessible via setlocale().
    
    This function converts the calling thread to use that global locale, if
    not already there.
```
The test failures look like this:
```
t/10_base.t ...... 
ok 1 - use Tk;
ok 2 - use Tk::Clock;
...
ok 33 -         Station clock: Time offset -4'05:06:07
ok 34 - Destroy Clock
XS_Tk__Callback_Call error:panic: Mismatch between what Perl thinks LC_CTYPE is (en_US.UTF-8) and what internal glibc thinks (C)

# Tests were run but no plan was declared and done_testing() was not seen.
Dubious, test returned 254 (wstat 65024, 0xfe00)
All 34 subtests passed 
...
t/30_dual.t ...... 
ok 1 - use Tk;
ok 2 - use Tk::Clock;
ok 18 - Destroy Clock
XS_Tk__Callback_Call error:panic: Mismatch between what Perl thinks LC_CTYPE is (en_US.UTF-8) and what internal glibc thinks (C)

# Tests were run but no plan was declared and done_testing() was not seen.
Dubious, test returned 254 (wstat 65024, 0xfe00)
All 18 subtests passed 
...
Test Summary Report
-------------------
t/10_base.t    (Wstat: 65024 Tests: 34 Failed: 0)
  Non-zero exit status: 254
  Parse errors: No plan found in TAP output
t/30_dual.t    (Wstat: 65024 Tests: 18 Failed: 0)
  Non-zero exit status: 254
  Parse errors: No plan found in TAP output
```
Although the problematic commit dates from Feb 2018, it apparently never generated errors until a new version of Tk-Clock was uploaded in Apr 2020 and subsequently tested with a variety of perls built with `-Dusethreads -DEBUGGING=both`.  Similar threaded perls built with `-DEBUGGING=-g` do not, AFAICT, exhibit these test failures.

Since this is not a regression in the 5.31 development cycle, this is not a blocker for the release of 5.32.0.  @Tux, @khwilliamson can you take a look?

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
See https://rt.cpan.org/Ticket/Display.html?id=132669#txn-1899534.

My thinking now is that this is a "downstream code must adapt to change in blead" situation, _i.e.,_ the change to perl back in 58e641fba in 2018 is fine but exposes sub-optimal code in a downstream module or program which needs inspection.  @khwilliamson @eserte @Tux can you take a look?

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> See https://rt.cpan.org/Ticket/Display.html?id=132669#txn-1899534.
> 
> My thinking now is that this is a "downstream code must adapt to change in blead" situation, _i.e.,_ the change to perl back in [58e641f](https://github.com/Perl/perl5/commit/58e641fba50d93bcd84c0dc042b36d201958293d) in 2018 is fine but exposes sub-optimal code in a downstream module or program which needs inspection. @khwilliamson @eserte @Tux can you take a look?
> 
> Thank you very much.
> Jim Keenan

This is under discussion in the [Tk-Clock bug tracker](https://rt.cpan.org/Ticket/Display.html?id=132669).

**Comment by jkeenan** [comment]
> > See https://rt.cpan.org/Ticket/Display.html?id=132669#txn-1899534.
> > My thinking now is that this is a "downstream code must adapt to change in blead" situation, _i.e.,_ the change to perl back in [58e641f](https://github.com/Perl/perl5/commit/58e641fba50d93bcd84c0dc042b36d201958293d) in 2018 is fine but exposes sub-optimal code in a downstream module or program which needs inspection. @khwilliamson @eserte @Tux can you take a look?
> > Thank you very much.
> > Jim Keenan
> 
> This is under discussion in the [Tk-Clock bug tracker](https://rt.cpan.org/Ticket/Display.html?id=132669).

That ticket was in turn migrated to https://github.com/Tux/Tk-Clock/issues/1; issue is still open.

**Comment by Tux** [comment]
> That ticket was in turn migrated to [Tux/Tk-Clock#1](https://github.com/Tux/Tk-Clock/issues/1); issue is still open.

@jkeenan True, and I still have no idea how to address it in plain perl. I could use some help here.
